animated-transformer/src/lib/transformer/gpt2.ts

```diff
@@ -0,0 +1,574 @@
+    inputToFF: inputToFF.rename('inputRepToFF', 'inputRep'),
+    seqOutput,
+  };
+}
+
+export function initDecoderParams(config: TransformerConfig): TransformerParams {
+  const { spec, init } = config;
+  // const paramInitializerConfig = config.init;
+  const layers: AttnHeadParams[] = spec.layers.map((layerSpec) => {
+    const attnHeadSpec: AttnHeadParamSpec = {
+      inputRep: spec.inputRep,
+      kq: spec.kqvRep,
+      heads: layerSpec.nHeads,
+      value: spec.kqvRep,
+      layerNormPreAttention: layerSpec.layerNormPreAttention,
+      layerNormHeadsProjection: layerSpec.layerNormHeadsProjection,
+      // addLayerNormBias: AttentionIsAllYouNeed = true; T5 = false.
+      addLayerNormBias: layerSpec.addLayerNormBias,
+    };
+    return initAttnHeadParams(attnHeadSpec, init);
+  });
+  const tokenEmbedding = makeTruncNormal(
+    {
+      tokenId: config.tokenRep.tokens.length,
+      inputRep: spec.inputRep,
+    },
+    init
+  );
+
+  let transformerParams: TransformerParams = {
+    tokenEmbedding: tokenEmbedding,
+    layers: layers
+  };
+
+  if (spec.addPosEmbeddings) {
+    const posEmbedding = makeTruncNormal({
+        posId: spec.posEncodingSeqLength,
+        inputRep: spec.inputRep,
+      });
+    transformerParams.posEmbedding = posEmbedding;
+  }
+
+  if (spec.layerNorm) {
+    const layerNormParams = initLayerNormParamsWithDims(
+        spec.addLayerNormBias, {'inputRep': spec.inputRep});
+    transformerParams.layerNorm = layerNormParams
+  }
+
+  return transformerParams;
+}
+
+export type TransformerComputation = {
+  layers: BatchAttnHeadCompututation[];
+};
+
+export function computeTransformer(
+  model: {
+    config: { spec: TransformerParamSpec };
+    params: TransformerParams;
+  },
+  seqInput: GTensor<'batch' | 'pos' | 'inputRep'>,
+  generator: RandomStream
+): TransformerComputation {
+  const compute: TransformerComputation = { layers: [] };
+  let currentLayerInput = dropout(model.config.spec.computeSpec.dropoutRate, 
+    seqInput, generator.random());
+  model.params.layers.forEach((layerParams, i) => {
+    const layerCompute = computeAttnHead(
+      model.config.spec.layers[i].computeSpec,
+      layerParams,
+      currentLayerInput,
+      generator
+    );
+    compute.layers.push(layerCompute);
+    currentLayerInput = layerCompute.seqOutput;
+  });
+
+  // TODO(@aliciafmachado): Hacky way to apply layer norm after the final layer.
+  if (model.params.layerNorm) {
+    const lastBlock = compute.layers[compute.layers.length - 1];
+    const finalSeqOutput = layerNorm(model.params.layerNorm, lastBlock.seqOutput, 'inputRep', 
+        model.config.spec.computeSpec.layerNormEpsilon);
+    let finalOutput = lastBlock;
+    finalOutput.seqOutput = finalSeqOutput;
+    finalOutput.seqInput = lastBlock.seqOutput;
+    compute.layers.push(finalOutput);
+  }
+  
+  return compute;
+}
+
+/** Batch compute the loss for the last token of a transformer.
```

Comment:
I think the comment here is wrong; this doesn't compute the loss but the final tokenId's for each example in the batch.

Reply:
Same as before (came with the file). Fixed the comment to say the right thing!